### PR TITLE
feat(tiller): support CRD installation

### PR DESCRIPTION
**Pull request by technosophos**
This adds support for installing CRDs well before any other resource
kinds are installed.

This PR introduces a new hook, `crd-install`, that fires before
manifests are even validated. It is used to install a CRD before any
other part of a chart is installed.

Currently, this hook is _only implemented for install_. That means we
currently cannot add new CRDs during `helm upgrade`, nor can they
be rolled back. This is the safest configuration, as the update/rollback
cycle gets very challenging when CRDs are added and removed.

### CODE REVIEW

**Comment by rimusz** [comment]
yay, long awaited feature 😀

**Comment by technosophos** [comment]
Okay, here's how this works

First, if you are _just_ creating a CRD with no instances of that CRD, you do nothing (just as always). This is designed to allow for a CRD and instances of the CRD to be installed in the same chart. It is also designed to be maximally backward compatible, though I did have to eliminate one set of validations in order to do this.

To add a CRD to a chart, create a CRD template and set an annotation on the template:

```yaml
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: crontabs.stable.example.com
  # THE IMPORTANT PART!!!
  annotations:
    "helm.sh/hook": crd-install
spec:
  group: stable.example.com
  version: v1
  scope: Namespaced
  names:
    plural: crontabs
    singular: crontab
    kind: CronTab
    shortNames:
    - ct
```

You can also create a template that references that CRD:

```yaml
apiVersion: stable.example.com/v1
kind: CronTab
metadata:
  name: {{ .Release.Name }}-inst
```

Now, when you run `helm install ./mychart`, the following will happen:

- Manifests will be sorted and rendered as usual
- `crd-install` hooks will be separated
- If dry-run and there's a CRD hook, _validation will be skipped and the user will be warned_
- Otherwise, execute the `crd-install` hook (install the CRD)
- Validate the remaining normal manifests
- Run `pre-install`
- Continue on the normal installation path

If `helm install --no-crd-hook` is specified, then the `crd-install` hook is skipped, and no CRDs are installed. If `--no-hooks` is specified, `crd-install` is skipped. If `--no-crd-hook --dry-run` are specified together, the validator will run.

File: pkg/tiller/hooks.go [hunk 1]
```diff
@@ -137,10 +138,6 @@ func (file *manifestFile) sort(result *result) error {
 			return e
 		}
 
-		if entry.Version != "" && !file.apis.Has(entry.Version) {
```
Comment (technosophos) [inline pkg/tiller/hooks.go]: To be really clear: I am removing this set of checks. It appears to duplicate the check in validateManifest, and it fires well before we have a chance to install the CRD.

**Review by adamreese — APPROVED** [review]

**Comment by tamalsaha** [comment]
@technosophos , what should I do if I want to support last 2-3 version of helm with the same chart?

**Comment by bamb00** [comment]
What helm version is this feature going to be in?

**Comment by bacongobbler** [comment]
2.10

**Comment by gyliu513** [comment]
What is the difference of this fix with https://github.com/helm/helm/pull/3019 ? @technosophos

**Comment by bacongobbler** [comment]
#3019 is for custom resources (CRs), whereas this PR adds support for custom resource definitions (CRDs). Prior to this hook being available, charts could not install CRDs.

The docs that were added to this PR goes into a little more detail on this new hook, which can be found here: https://docs.helm.sh/developing_charts/#defining-a-crd-with-the-crd-install-hook

Does that help explain this fix, or is there something else we should document?

**Comment by gyliu513** [comment]
@bacongobbler I think https://github.com/helm/helm/pull/3019 is also for CRD but not CR, https://github.com/helm/helm/pull/3019 is trying to enable all of the CRDs should be created first before other resources take reference of this CRD. Am I missing anything?

**Pull request merged**